test/common/buffer/buffer_fuzz_test.cc

```diff
@@ -0,0 +1,221 @@
+    uint32_t allocated_slices = target_buffer.reserve(reserve_length, slices, reserve_slices);
+    uint32_t allocated_length = 0;
+    for (uint32_t i = 0; i < allocated_slices; ++i) {
+      ::memset(slices[i].mem_, 'c', slices[i].len_);
+      allocated_length += slices[i].len_;
+    }
+    const uint32_t target_length =
+        std::min(allocated_length, action.reserve_commit().commit_length());
+    uint32_t shrink_length = allocated_length;
+    while (shrink_length > target_length) {
+      ASSERT(allocated_slices > 0);
+      const uint32_t available = slices[allocated_slices - 1].len_;
+      const uint32_t remainder = shrink_length - target_length;
+      if (available > remainder) {
+        slices[allocated_slices - 1].len_ -= remainder;
+        break;
+      }
+      shrink_length -= available;
+      --allocated_slices;
+    }
+    target_buffer.commit(slices, allocated_slices);
+    ASSERT_EQ(previous_length + target_length, target_buffer.length());
+    break;
+  }
+  case test::common::buffer::Action::kCopyOut: {
+    const uint32_t start = action.copy_out().start() % target_buffer.length();
+    uint8_t copy_buffer[2 * 1024 * 1024];
+    const uint32_t length =
+        std::min(static_cast<uint32_t>(target_buffer.length()),
+                 std::min(action.copy_out().length(), static_cast<uint32_t>(sizeof(copy_buffer))));
+    target_buffer.copyOut(start, length, copy_buffer);
+    const std::string contents = target_buffer.toString();
+    ASSERT_EQ(0, ::memcmp(copy_buffer, contents.data(), length));
+    break;
+  }
+  case test::common::buffer::Action::kDrain: {
+    const uint32_t previous_length = target_buffer.length();
+    const uint32_t drain_length =
+        std::min(static_cast<uint32_t>(target_buffer.length()), action.drain());
+    target_buffer.drain(drain_length);
+    ASSERT_EQ(previous_length - drain_length, target_buffer.length());
+    break;
+  }
+  case test::common::buffer::Action::kLinearize: {
+    target_buffer.linearize(
+        std::min(static_cast<uint32_t>(target_buffer.length()), action.linearize()));
+    break;
+  }
+  case test::common::buffer::Action::kMove: {
+    const uint32_t source_index = action.move().source_index() % BufferCount;
+    if (target_index == source_index) {
+      break;
+    }
+    Buffer::OwnedImpl& source_buffer = buffers[source_index];
+    if (action.move().length() == 0) {
+      target_buffer.move(source_buffer);
+    } else {
+      target_buffer.move(source_buffer, std::min(static_cast<uint32_t>(source_buffer.length()),
+                                                 action.move().length()));
+    }
+    break;
+  }
+  case test::common::buffer::Action::kRead: {
+    int pipe_fds[2] = {0, 0};
+    ASSERT_EQ(0, pipe(pipe_fds));
+    const uint32_t max_length = clampSize(action.read());
+    std::string data(max_length, 'd');
+    const uint32_t previous_length = target_buffer.length();
+    ASSERT_GT(::write(pipe_fds[1], data.data(), max_length), 0);
```

Comment:
I think it's possible for this code to deadlock if `max_length` exceeds whatever the pipe's internal buffer space is. The write will block, waiting for a reader to drain enough data from the pipe, but here there's nobody reading until after the write is fully complete.

Possible solutions that come to mind:
- Run the write and read concurrently (in separate threads, or in an event loop).
- Make the pipe nonblocking, keep track of how much data the write call succeeded in writing, and validate that the read picks up that much data.

Reply:
I think non-blocking is fine for these purposes.